Fix `findmin/max` with `dims` for non-1 based inputs.

N5N3 opened the pull request:
This PR fixs the possible error from `findmin/max` by ensuring that the input's `keys` contains no zero. (Close #38660)
The added check might be expensive (`zi in keys` calls `iterate` based fallback), but I can't find a non-breaking way to do the same check.

This PR also add a faster fallback if the input's `keys` is a `AbstractArray` with the same `axes`.
My local bench shows that it makes `findmin/max` almost 2x faster if `f` is cheap.

N5N3 committed:
Check the legality of zero-index based initialization.

And add an optimized fallback for `keys::AbstractArray` (almost 2x faster with cheap `f`.)

andrewjradcliffe reviewed (commented):
This is a good idea. Once approved, I'll need to update [#45783](https://github.com/JuliaLang/julia/pull/45783) to utilize the same approach.

N5N3 commented on base/reducedim.jl:
```diff
@@ -1070,7 +1074,57 @@ function findminmax!(f, op, Rval, Rind, A::AbstractArray{T,N}) where {T,N}
+        axes(Rval, i) == axes(Rind, i) || throw(DimensionMismatch("Find-reduction: outputs must have the same indices"))
+    end
+    if init
+        copyfirst!(Rind, keys)
+        mapfirst!(f, Rval, A)
```
Nice catch. Before this PR, the only usage for `mapfirst!` are the initialization for `maximum!`.
The redundancy also happens there as `Base._mapreducedim!` also performs `check_reducedims`.
I think it's ok to remove the `check_reducedims` within `mapfirst!` as it use `view` rather than `@inbounds view`.

N5N3 committed:
Adopt suggestions.

1. remove `check_reducedims` within `mapfirst!`.
    `init_array!` for other reductions also skip this check.
     it should be safe to remove it as we add no `@inbounds`
2. remove unneeded `getindex`.

Co-Authored-By: Andrew Radcliffe <96091198+andrewjradcliffe@users.noreply.github.com>

N5N3 commented on base/reducedim.jl:
```diff
@@ -1070,7 +1074,57 @@ function findminmax!(f, op, Rval, Rind, A::AbstractArray{T,N}) where {T,N}
+            for i in axes(A, 1)
+                tmpAv = f(A[i,IA])
+                tmpRv = Rval[i,IR]
+                tmpRi = Rind[i,IR]
```
Removed, thanks!

vtjnash committed:
Merge branch 'master' into findminmax
